### ISSUE TITLE
Watch the stream of service status changes

### ISSUE THREAD

**Issue by tmrts** [systemd/systemd#3726]
### Submission type
- [ ] Bug report
- [X] Request for enhancement (RFE)
### Feature

The ability to watch the status events via `systemctl status --watch` or `systemctl watch` API.

I would like to watch events which is not possible to do efficiently at the moment.

Currently a user can either:
1. `journalctl -f` and parse the logs manually to interpret events
2. Perform polling using `systemctl status` in an interval

cc @alban @alepuccetti @lucab

**Comment by jonboulle** [comment]
What events do you want more specifically? Much of this is already exposed over D-Bus. https://github.com/coreos/rkt/issues/1193#issuecomment-125399764

**Comment by alban** [comment]
I was thinking that "systemctl watch" could be implemented by calling Subscribe()`and receiving the D-Bus signals`UnitNew`and`UnitRemoved` and others from [the systemd Manager interface (org.freedesktop.systemd1.Manager)](https://www.freedesktop.org/wiki/Software/systemd/dbus/).

Example:

```
$ systemctl watch
busybox.service loaded
busybox.service started
busybox.service stopped
busybox.service unloaded
^C
$
```

If systemctl and systemd communicate via the systemd's private socket, then `systemctl watch` could work even when dbus-daemon is not installed.

**Comment by poettering** [comment]
Why isn't journalctl -f _PID=1 enough?

**Comment by alban** [comment]
If systemd is running in a container, should `journalctl -f _PID=${SYSTEMD_PID}` on the host work?

**Comment by iaguis** [comment]
It seems the PID recorded is `1` for systemd in the container.

**Comment by alban** [comment]
With `journalctl -m  _PID=1`, I see the following when I start rkt:

```
Jul 18 12:30:40 rainbow systemd[1]: Started Container rkt-9f85d9e9-be24-4984-a62d-bdb74fce314b.
Jul 18 12:33:34 rainbow systemd[1]: Started Container rkt-b5ea65fa-b64d-441f-a596-e20001128ca1.
Jul 18 12:33:35 rkt-b5ea65fa-b64d-441f-a596-e20001128ca1 systemd[1]: Starting Prepare minimum environment for chrooted applications...
Jul 18 12:33:35 rkt-b5ea65fa-b64d-441f-a596-e20001128ca1 systemd[1]: Started Prepare minimum environment for chrooted applications.
Jul 18 12:33:35 rkt-b5ea65fa-b64d-441f-a596-e20001128ca1 systemd[1]: Started Application=busybox Image=registry-1.docker.io/library/busybox.
Jul 18 12:33:35 rkt-b5ea65fa-b64d-441f-a596-e20001128ca1 systemd[1]: Reached target rkt apps target.
Jul 18 12:33:35 rkt-b5ea65fa-b64d-441f-a596-e20001128ca1 systemd[1]: Startup finished in 173ms.
```

So I see the description field of the units inside the container, but not their name. For example, `busybox.service` contains:

```
[Unit]
Description=Application=busybox Image=registry-1.docker.io/library/busybox
```

That's what I see in the log, but not "Starting busybox.service". Is there a way around this? It makes it more difficult to parse it automatically.

**Comment by poettering** [comment]
> That's what I see in the log, but not "Starting busybox.service". Is there a way around this? It makes it more difficult to parse it automatically.

if you use "journalctl -o verbose" you should see it, no? And anyway, you really shouldn't parse the output of "journalctl -o short" anyway, it's for human consumption. Use "journalctl -o json" or something...

**Comment by martinpitt** [comment]
I just filed #3830 which is similar indeed. This RFE asks for an (indefinitively running) monitoring while I need to wait until a given state is reached; but the subscription/event watching code would be the same.

I wonder if we could unify both into the same verb `monitor` or `watch` and determine whether or not to exit on a particular state depending on whether we have a `--state` option. I. e.

```
systemctl watch foo.service
```

would wait/monitor indefinitely, while

```
systemctl watch --state inactive foo.service
```

would exit as soon as the state is or changes to `inactive`. However, that would not stop on `failed`, so `--state` would need to become a list or interpret `failed` as `inactive` (but that's ugly special-casing).

**Comment by alban** [comment]
@martinpitt I agree it would be best if the same watching code was shared for both feature. This issue comes from https://github.com/coreos/rkt/issues/1193 and we would need to watch several services (typically, 1, 2 or a few).

Would it make sense to reuse the same verb for the following?

```
systemctl watch foo.service bar.service other.service
```

**Comment by martinpitt** [comment]
@poettering: WDYT about the proposal above? I. e. just `systemctl watch` for an indefinite monitoring, and `watch --state inactive,failed` for waiting until one of the given states are reached? This should suit both this one and issue #3830.

**Comment by poettering** [comment]
sure, sounds acceptable. any such work should be based on the Ref()/Unref() concept introduced with #3984, to make watching unit exit status race-free. Moreover, it should probably show a similar summary on exit as "systemd-run --wait" with that PR does.

**Comment by martinpitt** [comment]
I had a first look at this. It seems there is not actually that much code overlap as the two modes differ in pretty much every regard. `systemctl start --wait` is very similar to `systemd-run --wait` and just waits until the unit becomes inactive again (no further states necessary). It would subscribe to `PropertiesChanged` only on that one object path, to avoid a lot of unnecessary wakeups. It would also ref/unref the unit to avoid missing the signal.

OTOH a global `systemctl watch` would not take any unit names, run forever, and be a global operation; it needs to subscribe to `PropertiesChanged` without an object path and thus watch a whole lot more events (which is fine as this is for debugging while `start --wait` is intended for regular production operation). It also has to subscribe to `UnitNew` and similar signals, and would not ref/unref.

So I think it would be cleaner to keep these two apart.

**Comment by eMPee584** [comment]
(… so for the time being, `watch -c  SYSTEMD_URLIFY=0 SYSTEMD_COLORS=1 systemctl status foo.service` is an appropriate solution (c.f. https://github.com/systemd/systemd/issues/12356 for  why it needs `SYSTEMD_URLIFY=0`).)

**Comment by akvadrako** [comment]
It would be nice if `systemctl status --watch` also shows new log messages as they arrive.

**Comment by Simbi0z** [comment]
> (… so for the time being, `watch -c SYSTEMD_URLIFY=0 SYSTEMD_COLORS=1 systemctl status foo.service` is an appropriate solution (c.f. #12356 for why it needs `SYSTEMD_URLIFY=0`).)

This does not work appropriately in tmux. Any workarounds?
